ddl/job_table.go

```diff
@@ -0,0 +1,564 @@
+		return
+	}
+	d.mu.RLock()
+	d.mu.hook.OnGetJobAfter(pool.tp().String(), job)
+	d.mu.RUnlock()
+
+	d.delivery2worker(wk, pool, job)
+}
+
+func (d *ddl) delivery2worker(wk *worker, pool *workerPool, job *model.Job) {
+	injectFailPointForGetJob(job)
+	d.insertRunningDDLJobMap(job.ID)
+	d.wg.Run(func() {
+		metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Inc()
+		defer func() {
+			pool.put(wk)
+			d.deleteRunningDDLJobMap(job.ID)
+			asyncNotify(d.ddlJobCh)
+			metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Dec()
+		}()
+		// we should wait 2 * d.lease time to guarantee all TiDB server have finished the schema change.
+		// see waitSchemaSynced for more details.
+		if !d.isSynced(job) || d.once.Load() {
+			wk.waitSchemaSynced(d.ddlCtx, job, 2*d.lease)
+			d.once.Store(false)
+		}
+		if err := wk.HandleDDLJobTable(d.ddlCtx, job); err != nil {
+			logutil.BgLogger().Info("[ddl] handle ddl job failed", zap.Error(err), zap.String("job", job.String()))
+		}
+	})
+}
+
+func (d *ddl) markJobProcessing(sess *session, job *model.Job) error {
+	sess.SetDiskFullOpt(kvrpcpb.DiskFullOpt_AllowedOnAlmostFull)
+	_, err := sess.execute(context.Background(), fmt.Sprintf("update mysql.tidb_ddl_job set processing = 1 where job_id = %d", job.ID), "mark_job_processing")
+	return errors.Trace(err)
+}
+
+const (
+	addDDLJobSQL    = "insert into mysql.tidb_ddl_job(job_id, reorg, schema_ids, table_ids, job_meta, type, processing) values"
+	updateDDLJobSQL = "update mysql.tidb_ddl_job set job_meta = %s where job_id = %d"
+)
+
+func insertDDLJobs2Table(sess *session, jobs []*model.Job, updateRawArgs bool) error {
+	failpoint.Inject("mockAddBatchDDLJobsErr", func(val failpoint.Value) {
+		if val.(bool) {
+			failpoint.Return(errors.Errorf("mockAddBatchDDLJobsErr"))
+		}
+	})
+	if len(jobs) == 0 {
+		return nil
+	}
+	var sql bytes.Buffer
+	sql.WriteString(addDDLJobSQL)
+	for i, job := range jobs {
+		b, err := job.Encode(updateRawArgs)
+		if err != nil {
+			return err
+		}
+		if i != 0 {
+			sql.WriteString(",")
+		}
+		sql.WriteString(fmt.Sprintf("(%d, %t, %s, %s, %s, %d, %t)", job.ID, job.MayNeedReorg(), strconv.Quote(job2SchemaIDs(job)), strconv.Quote(job2TableIDs(job)), wrapKey2String(b), job.Type, !job.NotStarted()))
+	}
+	sess.SetDiskFullOpt(kvrpcpb.DiskFullOpt_AllowedOnAlmostFull)
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnDDL)
+	_, err := sess.execute(ctx, sql.String(), "insert_job")
+	logutil.BgLogger().Debug("[ddl] add job to mysql.tidb_ddl_job table", zap.String("sql", sql.String()))
+	return errors.Trace(err)
+}
+
+func job2SchemaIDs(job *model.Job) string {
+	return job2UniqueIDs(job, true)
+}
+
+func job2TableIDs(job *model.Job) string {
+	return job2UniqueIDs(job, false)
+}
+
+func job2UniqueIDs(job *model.Job, schema bool) string {
+	switch job.Type {
+	case model.ActionExchangeTablePartition, model.ActionRenameTables:
+		var ids []int64
+		if schema {
+			ids = job.CtxVars[0].([]int64)
```

Comment:
For `ActionRenameTables`, it seams only add old schema, how about the new schema(sometimes old schema ID != new schema ID)?

Reply:
they are all added, you can see how the `renameTables` job build